File: common/src/main/java/juuxel/adorn/CommonEventHandlers.java
package juuxel.adorn;

import juuxel.adorn.block.CarpetedBlock;
import net.minecraft.world.level.block.WoolCarpetBlock;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.BlockItem;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.level.Level;

public final class CommonEventHandlers {
    public static InteractionResult handleCarpets(Player player, Level world, InteractionHand hand, BlockHitResult hit) {
        var stack = player.getItemInHand(hand);

        if (stack.getItem() instanceof BlockItem blockItem) {
            if (blockItem.getBlock() instanceof WoolCarpetBlock carpet) {
                var pos = hit.getBlockPos().relative(hit.getDirection());
                var state = world.getBlockState(pos);

                if (state.getBlock() instanceof CarpetedBlock carpeted && carpeted.canStateBeCarpeted(state)) {
                    world.setBlockAndUpdate(pos, state.setValue(CarpetedBlock.CARPET, CarpetedBlock.CARPET.wrapOrNone(carpet.getColor())));
                    var soundGroup = carpet.defaultBlockState().getSoundType();
                    world.playSound(
                        player, pos, soundGroup.getPlaceSound(), SoundSource.BLOCKS,
                        (soundGroup.volume + 1f) * 0.5f, soundGroup.pitch * 0.8f
                    );

                    if (!player.getAbilities().instabuild) {
                        stack.shrink(1);
                    }
                    player.swing(hand);
                    return InteractionResult.SUCCESS;
                }
            }
        }

        return InteractionResult.PASS;
    }
}

File: common/src/main/kotlin/juuxel/adorn/CommonEventHandlers.kt
package juuxel.adorn

import juuxel.adorn.block.AdornBlockEntities
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.util.math.BlockPos

object CommonEventHandlers {
    // TODO: This currently targets block attacking. What about explosions and mining machines?
    fun shouldCancelBlockBreak(player: PlayerEntity, pos: BlockPos): Boolean {
        if (player.isSpectator || player.isCreativeLevelTwoOp) return false

        val tradingStation = player.world.getBlockEntity(pos, AdornBlockEntities.TRADING_STATION).orElse(null) ?: return false
        return !tradingStation.isOwner(player)
    }
}
